software/contrib/master_clock.py

```diff
@@ -0,0 +1,320 @@
+                if self.previousSelectedDivision != selectedDivision:
+                    self.outputDivisions[self.activeOption - 3] = selectedDivision
+                
+                self.previousSelectedDivision = selectedDivision
+            
+            self.previousActiveOption = self.activeOption
+                    
+        else:
+            configMarker = '.'
+        
+        oled.text(str(self.bpm) + ' bpm', 6, 0, 1)
+        oled.text(str(self.pulseWidthPercent) + ':' + str(str(self.pulseWidthMs)), 75, 0, 1)
+        oled.text('/' + str(self.outputDivisions[0]), 6, 12, 1)
+        oled.text('/' + str(self.outputDivisions[1]), 45, 12, 1)
+        oled.text('/' + str(self.outputDivisions[2]), 85, 12, 1)
+        oled.text('/' + str(self.outputDivisions[3]), 6, 24, 1)
+        oled.text('/' + str(self.outputDivisions[4]), 45, 24, 1)
+        oled.text('/' + str(self.outputDivisions[5]), 85, 24, 1)
+        oled.text(configMarker, self.markerPositions[self.activeOption-1][0], self.markerPositions[self.activeOption-1][1], 1)
+        oled.show() 
+
+    ''' Holds given output (cv) high for pulseWidthMs duration '''
+    async def outputPulse(self, cv):
+        cv.voltage(5)
+        await asyncio.sleep_ms(self.pulseWidthMs)
+        cv.off()
+
+    ''' Given a desired BPM, calculate the time to sleep between clock pulses '''
+    def calcSleepTime(self):
+        self.timeToSleepMs = int((60000 / self.bpm / self.CLOCKS_PER_QUARTER_NOTE))
+    
+    def checkForAinBPM(self):
+        val = 100 * ain.percent()
+        # If there is an analogue input voltage use that for BPM. clamp ensures it is higher than MIN and lower than MAX
+        if val > self.minAnalogInputVoltage:
+            self.bpm = clamp(int((((self.MAX_BPM) / 100) * val) + self.MIN_BPM), self.MIN_BPM, self.MAX_BPM)
+            self.calcSleepTime()
+            self.getPulseWidth()
+        else:
+            # No analog input, revert to last saved state
+            self.bpm = self.state.get("bpm", 100)
+            self.calcSleepTime()
+            self.getPulseWidth()
+
+    def getPulseWidth(self):
+        # Set max to 50% of total cycle time
+        self.MAX_PULSE_WIDTH = self.timeToSleepMs // 2
+        # Get desired PW percent. clamp ensures it is higher than MIN and lower than MAX
+        self.pulseWidthMs = clamp(int((self.MAX_PULSE_WIDTH * 2) * (self.pulseWidthPercent / 100)), self.MIN_PULSE_WIDTH, self.MAX_PULSE_WIDTH)
+
+    ''' Triggered by main. Sends output pulses at required division '''
+    def clockTrigger(self):
+
+        if self.DEBUG:
+            print('BPM: ' + str(self.bpm) + ' cycle: ' + str(self.timeToSleepMs) + ' PW:' + str(self.pulseWidthMs))
+
+        for idx, output in enumerate(self.outputDivisions):
+            if output != 0:
+                if self.step % output == 0:
+                    self.el.create_task(self.outputPulse(cvs[idx]))
```

Comment:
[discussion] I didn't catch this the first time I read over this code, but I suspect this is the source of the wonky playback in config mode. It's creating a new async task for all 6 outputs on each clock pulse. When BPM changes, the existing tasks are not killed and therefore still running with their previous values. This is also similar to the work I was doing with my [varigate clone](https://github.com/awonak/EuroPi/blob/gates/software/contrib/gates.py) which had similar challenges with accurate async task execution. In my case, the problem was retriggering, which caused too many async events simultaneously and killed accuracy. I imagine this script will experience the same problem at higher BPM with several close clock divisions. 

Side note, I couldn't figure out how to save BMP changes to accurately test my hypothesis.

Reply:
BPM changes are auto-saved whenever you exit config mode (long b2 press). You can also exit config mode by changing screen (short b2 press).

Regarding the still running tasks, I will try and devise a test to get more info.

 In my current level of understanding, each new task simply holds the output open, (async) sleeps, then exits.
 However, leave this with me and I will find out more.

Reply:
I added some debug code and I believe I have concluded that all tasks successfully complete and no zombie tasks are left running in the background. See below for the code to test this. In my testing I did not see any times when a previous task was running. I'd appreciate a second pair of eyes on my code though and any more suggestions you may have.

Create a list to hold each asyncio.Task class object when it is created:
```
        self.tasks = []
        for n in range(6):
            self.tasks.append(0)
```

Add each asyncio.Task object into the list when creating a task:
`self.tasks[idx] = asyncio.create_task(self.outputPulse(cvs[idx]))`

If a new task will be created, but the previous one is not done print an error message:

```
                    if self.tasks[idx] != 0 and not self.tasks[idx].done():
                        print(f'Task: {idx} is not done')
```